### DISCUSSION TITLE
PNA without edge_attributes

### DISCUSSION THREAD

**Discussion by AdarshMJ** [pyg-team/pytorch_geometric#4470] [Q&A] [answered]
Hi! So Im trying to use - https://github.com/pyg-team/pytorch_geometric/blob/master/examples/pna.py on a custom dataset, but my dataset doesn't have edge_attr. Is there a way to generate dummy edge_attr to use it with the code ? Any suggestions would be welcome. Thank you in advance!

**Comment by rusty1s** [comment, marked as answer]
No need to create dummy edge features. Simply drop the edge_dim argument and do not pass edge_attr to the forward call of the GNN layer.

**Comment by AdarshMJ** [comment]
Hi! Thank you for the response. It works. But I was having a runtime error. Im not sure how to resolve this. Could you please point me in the direction as to where could be the problem. Thanks

![Screenshot 2022-04-14 at 01 21 13](https://user-images.githubusercontent.com/24205200/163285340-fd3721ab-be92-4127-aef3-d73e923654a7.png)

**Reply by rusty1s** [reply]
How does your data look like? What‘s the output of `print(data)` and `print(data.edge_index.max())`?

**Reply by AdarshMJ** [reply]
Hi!
It looks like this -
`data DataBatch(x=[655, 11], edge_index=[2, 1220], edge_attr=[1220, 4], y=[64, 19], pos=[655, 3], idx=[64], name=[64], z=[655], batch=[655], ptr=[65])
EDGEINDEX tensor(654)`

**Reply by rusty1s** [reply]
Mh, this looks correct. Does the problem also exist for other GNN layers? Can you show me the full traceback?

**Comment by AdarshMJ** [comment]
`data DataBatch(x=[646, 11], edge_index=[2, 1202], edge_attr=[1202, 4], y=[64, 19], pos=[646, 3], idx=[64], name=[64], z=[646], batch=[646], ptr=[65])
EDGEINDEX tensor(645)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-28-bd72154f131c>](https://localhost:8080/#) in <module>()
      1 for epoch in range(1, 301):
----> 2     loss = train(epoch)
      3    # val_mae = test(test_loader)
      4     test_mae = test(test_loader)
      5     scheduler.step(test_mae)

4 frames
[<ipython-input-27-647577627c35>](https://localhost:8080/#) in train(epoch)
      8         print("EDGEINDEX",data.edge_index.max())
      9         optimizer.zero_grad()
---> 10         out = model(data.x, data.edge_index, data.batch)
     11         loss = (out.squeeze() - data.y).abs().mean()
     12         loss.backward()

[/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in _call_impl(self, *input, **kwargs)
   1100         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1101                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1102             return forward_call(*input, **kwargs)
   1103         # Do not call functions when jit is used
   1104         full_backward_hooks, non_full_backward_hooks = [], []

[<ipython-input-26-0a3f51699588>](https://localhost:8080/#) in forward(self, x, edge_index, batch)
     49         x = self.node_emb(x.squeeze().long())
     50         for conv, batch_norm in zip(self.convs, self.batch_norms):
---> 51             x = F.relu(batch_norm(conv(x, edge_index)))
     52 
     53         x = global_add_pool(x, batch)

[/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in _call_impl(self, *input, **kwargs)
   1100         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1101                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1102             return forward_call(*input, **kwargs)
   1103         # Do not call functions when jit is used
   1104         full_backward_hooks, non_full_backward_hooks = [], []

[/usr/local/lib/python3.7/dist-packages/torch_geometric/nn/conv/pna_conv.py](https://localhost:8080/#) in forward(self, x, edge_index, edge_attr)
    153             x = x.view(-1, self.towers, self.F_in)
    154         else:
--> 155             x = x.view(-1, 1, self.F_in).repeat(1, self.towers, 1)
    156 
    157         # propagate_type: (x: Tensor, edge_attr: OptTensor)

RuntimeError: shape '[-1, 1, 75]' is invalid for input of size 177650`

**Reply by AdarshMJ** [reply]
It keeps alternating between this error and the one I posted earlier

**Reply by rusty1s** [reply]
Can you check with towers=1? I think in_channels need to be divisible by the number of towers.

**Reply by AdarshMJ** [reply]
No its still the same error.

**Reply by AdarshMJ** [reply]
`class Net(torch.nn.Module):
    def __init__(self):
        super().__init__()

        self.node_emb = Embedding(75, 25)
        #self.edge_emb = Embedding(4, 50)

        aggregators = ['mean', 'min', 'max', 'std']
        scalers = ['identity', 'amplification', 'attenuation']

        self.convs = ModuleList()
        self.batch_norms = ModuleList()
        for _ in range(4):
            conv = PNAConv(in_channels=75, out_channels=75,
                           aggregators=aggregators, scalers=scalers, deg=deg, towers=1, pre_layers=1, post_layers=1,
                           divide_input=False)
            self.convs.append(conv)
            self.batch_norms.append(BatchNorm(75))

        self.mlp = Sequential(Linear(75, 50), ReLU(), Linear(50, 25), ReLU(),
                              Linear(25, 1))

    def forward(self, x, edge_index, batch):

        x = self.node_emb(x.squeeze().long())
        for conv, batch_norm in zip(self.convs, self.batch_norms):
            x = F.relu(batch_norm(conv(x, edge_index)))

        x = global_add_pool(x, batch)
        return self.mlp(x)`

**Reply by AdarshMJ** [reply]
For this error - RuntimeError: The expanded size of the tensor (7271) must match the existing size (661) at non-singleton dimension 0.  Target sizes: [7271, 75].  Tensor sizes: [661, 1]
The problem seems to be - Input data is getting multiplied by 11, thats why 7271?

**Reply by rusty1s** [reply]
Ah, keep in mind that you should set the in_channels of the first layer to the feature size of your nodes. It currently is set to 75 but should be set to 11.

**Reply by AdarshMJ** [reply]
Should the self.node_emb change as well? 
If i keep in_channels = 11, i get this - RuntimeError: running_mean should contain 11 elements not 75, okay i think its because of batchnorm(75), if i change that to 11, we are back at this error -

 RuntimeError: The expanded size of the tensor (7271) must match the existing size (661) at non-singleton dimension 0. Target sizes: [7271, 75]. Tensor sizes: [661, 1]

**Reply by rusty1s** [reply]
You do not need the embedding layer in case your features are already of type floating point. This might also explain the shape mismatch. Simply drop it :)

**Reply by AdarshMJ** [reply]
Yes thank you. It works!
But I think I have another problem now, the loss is like really high , should I play around with the hyperparams to see what works or is there something else thats seriously wrong?

![Screenshot 2022-04-14 at 11 14 18](https://user-images.githubusercontent.com/24205200/163353899-849467f1-fae8-489d-aef5-43e96044b213.png)

**Reply by rusty1s** [reply]
It looks like there is something wrong with the data or the loss function. Is that a regression problem? Try to normalize your ground truth labels first.

**Reply by AdarshMJ** [reply]
Yes it’s a regression problem. I’m using the QM9 dataset currently.

**Reply by rusty1s** [reply]
Are you training against a single target? You should consider normalizing the targets as in `examples/am9_nn_conv.py`, see [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/qm9_nn_conv.py#L54).

**Reply by AdarshMJ** [reply]
Yes Im trying to train it against target 0. Im normalising as follows -  

```Python
mean = dataset.data.y[:, target].mean().item()
std = dataset.data.y[:, target].std().item()
dataset.data.y[:, target] = (dataset.data.y[:, target] - mean) / std
error += (model(data) * std - data.y * std).abs().sum().item()  # MAE```

**Reply by AdarshMJ** [reply]
But the MAE is still high
![Screenshot 2022-04-14 at 19 17 45](https://user-images.githubusercontent.com/24205200/163440421-aebde014-579d-4e6d-b65d-297e831f5ee1.png)

**Reply by rusty1s** [reply]
Thanks for confirming. It looks like something different is causing this. How does your final model architecture look like? Did you try to tune hyperparams such as the learning rate?

**Reply by AdarshMJ** [reply]
Hi Yes, I tried a couple of LRs, my final architecture looks like this -

```Python
class Net(torch.nn.Module):
    def __init__(self):
        super().__init__()

        #self.node_emb = Embedding(21, 11)
        #self.edge_emb = Embedding(4, 50)

        aggregators = ['mean', 'min', 'max', 'std']
        scalers = ['identity', 'amplification', 'attenuation']

        self.convs = ModuleList()
        self.batch_norms = ModuleList()
        for _ in range(4):
            conv = PNAConv(in_channels=dataset.num_features, out_channels=11,
                           aggregators=aggregators, scalers=scalers, deg=deg,
                           towers=11, pre_layers=1, post_layers=1,
                           divide_input=False)
            self.convs.append(conv)
            self.batch_norms.append(BatchNorm(11))

        self.mlp = Sequential(Linear(11, 64), ReLU(),Linear(64, 19))

    def forward(self, x, edge_index, batch):
        #x = self.node_emb(x.squeeze())
        #edge_attr = self.edge_emb(edge_attr)

        for conv, batch_norm in zip(self.convs, self.batch_norms):
            x = F.relu(batch_norm(conv(x, edge_index)))

        #x = global_add_pool(x, batch)
        x = global_add_pool(x, batch)
        return self.mlp(x)```

**Reply by AdarshMJ** [reply]
And changing the MAE loss to use mean instead of sum gives -
`total_error += (out.squeeze()*std - data.y*std).abs().mean().item()` 


![Screenshot 2022-04-15 at 19 14 10](https://user-images.githubusercontent.com/24205200/163600521-35884259-54ab-4428-aa56-84b827bf3076.png)

**Comment by rusty1s** [comment]
Note that your loss should be averaged across all examples (otherwise the loss value gets increased if the size of your dataset gets bigger). Currently, it looks like you are only averaging within mini-batches.

Your model architecture looks good to me, but I would probably start benchmarking with towers=1 and a reasonable hidden_channels value, e.g., 64 or 128.

**Reply by AdarshMJ** [reply]
Im sorry I dont fully understand,

 in the train function, there is this line which means its getting averaged over all the examples right?
`total_loss += loss.item() * data.num_graphs`
But in the test function, the loss calculated has an option to either use the 
`total_error += (model(data) * std - data.y * std).abs().sum().item()  # MAE`
or
`total_error += (model(data) * std - data.y * std).abs().mean().item()  # MAE`

If I use mean, the values are in the order of 1.78 and if I use sum, the values are really large like 300 etc.

**Reply by rusty1s** [reply]
Using sum is correct here. At the end, you need to divide by the number of graphs.

**Reply by AdarshMJ** [reply]
Thank you :)
